### PR TITLE
fix bad Frac(sparse polynomial ring over finite field)

### LINKED ISSUE #37374: Corrupted polynomial (wrong results/exceptions, crash SIGABRT) obtained from sparse FractionField (FLINT)
### Steps To Reproduce

The following code creates a univariate polynomial $f(x)$ modulo 37, which has very strange and wrong behavior. It misses some attributes (e.g. `factor(f)` can not find `.modulus`, `f.radical()` produces wrong result, wrong arithmetic, probably more).

```py
print(version())  
x = PolynomialRing(GF(37), ['x'], sparse=True).fraction_field().gen()
f = (x^8 + 16*x^6 + 4*x^4 + x^2 + 12).numerator()
print(type(f), f.parent())
print("  f", f)
print("f+1", f+1)
print("rad", f.radical())
print("r^2", f.radical()**2)
print(factor(f))
```
Output:
```
SageMath version 10.2, Release Date: 2023-12-03
<class 'sage.rings.polynomial.polynomial_zmod_flint.Polynomial_zmod_flint'> Sparse Univariate Polynomial Ring in x over Finite Field of size 37
  f x^8 + 16*x^6 + 4*x^4 + x^2 + 12
f+1 x^8 + 16*x^6 + 4*x^4 + x^2 + 12
rad x^4 + 15*x^2 + 26
r^2 x^8 + x^4
...
AttributeError: 'PolynomialRing_field_with_category' object has no attribute 'modulus'
```

While the actual radical should be just the square root of the polynomial which is:

```py
sage: x = GF(37)['x'].gen()
sage: f = (x^8 + 16*x^6 + 4*x^4 + x^2 + 12)
sage: f.radical()
x^4 + 8*x^2 + 7
sage: f.radical()**2
x^8 + 16*x^6 + 4*x^4 + x^2 + 12
sage: factor(f)
(x + 6)^2 * (x + 17)^2 * (x + 20)^2 * (x + 31)^2
```

### Expected Behavior

The output object is well formed and performs correctly (e.g. factor(f) succeeds, f.radical() is computed correctly, f+1 is computed correctly, etc.).

### Actual Behavior

The output object is corrupted and yields wrong results / exceptions.

### Additional Information

It might be that flint does not support sparse polynomials, while Sage wrongfully forces a dense object to be interpreted as a sparse one?

```py
sage: PolynomialRing(GF(37), ['x'], sparse=False, implementation="FLINT")
Univariate Polynomial Ring in x over Finite Field of size 37
sage: PolynomialRing(GF(37), ['x'], sparse=True, implementation="FLINT")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
...
ValueError: unknown implementation 'FLINT' for sparse polynomial rings over Finite Field of size 37
```

### Environment

```markdown
- **OS**: Linux Mint 21.3
- **Sage Version**: 10.2
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### PR DESCRIPTION
**Pull request by mezzarobba**
In passing, add support for fraction fields of polynomial rings over prime fields based on NTL.

Fixes #37374

### CODE REVIEW

File: src/sage/rings/polynomial/polynomial_ring.py [hunk 1]
```diff
@@ -2483,17 +2482,14 @@ def fraction_field(self):
             sage: t(x)
             x
 
+        Issue :issue:`37374`:
```
Comment (tscrim) [inline src/sage/rings/polynomial/polynomial_ring.py]:
```suggestion
        Fixed :issue:`37374`::
```

File: src/sage/rings/polynomial/polynomial_ring.py [hunk 2]
```diff
@@ -3566,6 +3562,31 @@ def irreducible_element(self, n, algorithm=None):
         # No suitable algorithm found, try algorithms from the base class.
         return PolynomialRing_dense_finite_field.irreducible_element(self, n, algorithm)
 
+    @cached_method
+    def fraction_field(self):
+        """
+        Returns the fraction field of self.
```
Comment (tscrim) [inline src/sage/rings/polynomial/polynomial_ring.py]:
```suggestion
        """
        Return the fraction field of ``self``.
```

File: src/sage/rings/fraction_field_FpT.pyx [hunk 3]
```diff
@@ -44,16 +44,26 @@ class FpT(FractionField_1poly_field):
         """
         INPUT:
 
-        - ``R`` -- A polynomial ring over a finite field of prime order `p` with `2 < p < 2^16`
+        - ``R`` -- A dense polynomial ring over a finite field of prime order `p` with `2 < p < 2^16`
```
Comment (tscrim) [inline src/sage/rings/fraction_field_FpT.pyx]:
```suggestion
        - ``R`` -- a dense polynomial ring over a finite field
          of prime order `p` with `2 < p < 2^16`
```
Two while-we-are-at-it changes.

File: src/sage/rings/polynomial/polynomial_modn_dense_ntl.pyx [hunk 4]
```diff
@@ -872,6 +864,21 @@ cdef class Polynomial_dense_modn_ntl_zz(Polynomial_dense_mod_n):
             sage: type(R(0)^0) == type(R(0))
             True
 
+        Negative powers work (over prime fields) but use a generic
```
Comment (tscrim) [inline src/sage/rings/polynomial/polynomial_modn_dense_ntl.pyx]:
```suggestion
        Negative powers work (over prime fields) but use the generic
```

**Review by tscrim — CHANGES_REQUESTED** [review]
Just a few little details.

**Comment by mezzarobba** [comment]
Thanks for your comments!

-- 
Marc

File: src/sage/rings/polynomial/polynomial_ring.py [hunk 5]
```diff
@@ -2483,17 +2482,14 @@ def fraction_field(self):
             sage: t(x)
             x
 
+        Issue :issue:`37374`::
```
Comment (tscrim) [inline src/sage/rings/polynomial/polynomial_ring.py]:
```suggestion
        :issue:`37374`::
```
This is redundant, including in the compiled doc.

File: src/sage/rings/fraction_field_FpT.pyx [hunk 6]
```diff
@@ -44,16 +44,27 @@ class FpT(FractionField_1poly_field):
         """
         INPUT:
 
-        - ``R`` -- A polynomial ring over a finite field of prime order `p` with `2 < p < 2^16`
+        - ``R`` -- a dense polynomial ring over a finite field of prime order
+          `p` with `2 < p < 2^16`
```
Comment (tscrim) [inline src/sage/rings/fraction_field_FpT.pyx]:
```suggestion
          `p` with `2 < p < 2^{16}`
```
Sorry, one more little thing to display properly.

**Review by tscrim — APPROVED** [review]
Thanks. You can set a positive review after these are changed.

**Comment by tscrim** [comment]
No problem. Thank you.

**Comment by vbraun** [comment]
```
sage -t --warn-long 38.3 --random-seed=123 src/sage/rings/polynomial/polynomial_modn_dense_ntl.pyx
**********************************************************************
File "src/sage/rings/polynomial/polynomial_modn_dense_ntl.pyx", line 878, in sage.rings.polynomial.polynomial_modn_dense_ntl.Polynomial_dense_modn_ntl_zz.__pow__
Failed example:
    (x-1)^(-5)
Expected:
    Traceback (most recent call last):
    ...
    TypeError: self must be an integral domain.
Got:
    <BLANKLINE>
    Traceback (most recent call last):
      File "sage/structure/category_object.pyx", line 855, in sage.structure.category_object.CategoryObject.getattr_from_category
        return self._cached_methods[name]
    KeyError: 'fraction_field'
    <BLANKLINE>
    During handling of the above exception, another exception occurred:
    <BLANKLINE>
    Traceback (most recent call last):
      File "sage/structure/element.pyx", line 2744, in sage.structure.element.RingElement._div_
        frac = self._parent.fraction_field()
      File "sage/structure/category_object.pyx", line 849, in sage.structure.category_object.CategoryObject.__getattr__
        return self.getattr_from_category(name)
      File "sage/structure/category_object.pyx", line 864, in sage.structure.category_object.CategoryObject.getattr_from_category
        attr = getattr_from_other_class(self, cls, name)
      File "sage/cpython/getattr.pyx", line 357, in sage.cpython.getattr.getattr_from_other_class
        raise AttributeError(dummy_error_message)
    AttributeError: 'PolynomialRing_dense_mod_n_with_category' object has no attribute '_SageObject__custom_name'
    <BLANKLINE>
    During handling of the above exception, another exception occurred:
    <BLANKLINE>
    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/doctest/forker.py", line 712, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/release/Sage/src/sage/doctest/forker.py", line 1147, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.rings.polynomial.polynomial_modn_dense_ntl.Polynomial_dense_modn_ntl_zz.__pow__[11]>", line 1, in <module>
        (x-Integer(1))**(-Integer(5))
        ~~~~~~~~~~~~~^^~~~~~~~~~~~~~~
      File "sage/rings/polynomial/polynomial_modn_dense_ntl.pyx", line 917, in sage.rings.polynomial.polynomial_modn_dense_ntl.Polynomial_dense_modn_ntl_zz.__pow__
        return ~r
      File "sage/rings/polynomial/polynomial_element.pyx", line 1522, in sage.rings.polynomial.polynomial_element.Polynomial.__invert__
        return self.parent().one() / self
      File "sage/rings/polynomial/polynomial_element.pyx", line 2701, in sage.rings.polynomial.polynomial_element.Polynomial.__truediv__
        return (<Element>left)._div_(right)
      File "sage/structure/element.pyx", line 2746, in sage.structure.element.RingElement._div_
        raise bin_op_exception('/', self, other)
    TypeError: unsupported operand parent(s) for /: 'Univariate Polynomial Ring in x over Ring of integers modulo 100 (using NTL)' and 'Univariate Polynomial Ring in x over Ring of integers modulo 100 (using NTL)'
**********************************************************************
1 item had failures:
   1 of  13 in sage.rings.polynomial.polynomial_modn_dense_ntl.Polynomial_dense_modn_ntl_zz.__pow__
    [335 tests, 1 failure, 0.31 s]
----------------------------------------------------------------------
sage -t --warn-long 38.3 --random-seed=123 src/sage/rings/polynomial/polynomial_modn_dense_ntl.pyx  # 1 doctest failed
----------------------------------------------------------------------
```

**Comment by mezzarobba** [comment]
> File "src/sage/rings/polynomial/polynomial_modn_dense_ntl.pyx", line 878, in sage.rings.polynomial.polynomial_modn_dense_ntl.Polynomial_dense_modn_ntl_zz.__pow__
> Failed example:
>     (x-1)^(-5)

This is with some additional commits merged in, right? Because I can't reproduce it with my branch alone. Anyway, I changed the test to ignore the exception test.

**Comment by vbraun** [comment]
Merge conflict

**Pull request merged**